generator/templates/structs.cpp.j2

```diff
@@ -0,0 +1,167 @@
+{%- endmacro -%}
+
+{%- macro get_message_type(bus, msg, msg_names) -%}
+{%- if msg.frame is defined -%}
+    {{bus.name}}::Message_Type_T::UNKNOWN_MSG,
+    {%- for sub_frame in msg.frame -%}
+        {{ get_message_type(bus, sub_frame, msg_names + [sub_frame.name]) }}
+    {%- endfor -%}
+{%- else -%}
+     Message_Type_T::{{'_'.join(msg_names)}},
+{%- endif -%}
+{%- endmacro -%}
+
+{%- macro get_messages(msg, msg_names) -%}
+{%- if msg.frame is defined -%}
+    nullptr,
+    {%- for sub_frame in msg.frame -%}
+        {{ get_messages(sub_frame, msg_names + [sub_frame.name]) }}
+    {%- endfor -%}
+{%- else -%}
+     &{{'::'.join(msg_names)}}_input,
+{%- endif -%}
+{%- endmacro -%}
+
+{%- macro get_keys(msg, msg_names) -%}
+{%- if msg.frame is defined -%}
+    {{ "::".join(msg_names) }}::key,
+    {%- for sub_frame in msg.frame -%}
+        {{ get_keys(sub_frame, msg_names + [sub_frame.name]) }}
+    {%- endfor -%}
+{%- else -%}
+     {{"::".join(msg_names)}}_T::key,
+{%- endif -%}
+{%- endmacro -%}
+
+{%- macro get_message_ranges(msg, msg_names) -%}
+{%- if msg.frame is defined -%}
+    std::make_pair({{ msg.slice.start }}, {{ msg.slice.length }}),
+    {%- for sub_frame in msg.frame -%}
+        {{ get_message_ranges(sub_frame, msg_names + [sub_frame.name]) }}
+    {%- endfor -%}
+{%- else -%}
+     std::make_pair(1, 0),
+{%- endif -%}
+{%- endmacro -%}
+
+{%- macro get_frame_ranges(msg, msg_names, cnt = 0) -%}
+{%- if msg.frame is defined -%}
+    {{ get_msg_len(msg) }},
+    {%- for sub_frame in msg.frame -%}
+        {{ get_frame_ranges(sub_frame, msg_names + [sub_frame.name], cnt + 1) }}
+    {%- endfor -%}
+{%- else -%}
+     0,
+{%- endif -%}
+{%- endmacro -%}
+
+#include <stdint.h>
+#include <stdbool.h>
+#include <chrono>
+#include "static.hpp"
+#include "bus.hpp"
+#include "structs.hpp"
+
+namespace CANlib {
+
+void send(Clock::time_point &last_sent_, Message_T* msg) {
+    if (Clock::now() - last_sent_ > msg->get_period()) {
+        last_sent_ = Clock::now();
+        Frame frame;
+        msg->unpack(frame);
+        TransmitFrame(frame, msg->get_bus_name());
+    }
+}
+
+void send(Message_T* msg) {
+    Frame frame;
+    msg->unpack(frame);
+    TransmitFrame(frame, msg->get_bus_name());
+}
+
+{%- for bus in can.bus %}
+namespace {{bus.name}} {
+{% for msg in bus.frame -%}
+{{ struct_declarations(bus, msg, msg.name) }}
+{%- endfor -%}
+
+extern const Message_Type_T message_types[{{get_msg_len(bus)}}] = {
+    Message_Type_T::UNKNOWN_MSG,
+    {% for msg in bus.frame -%}
+        {{ get_message_type(bus, msg, [msg.name]) }}
+    {%- endfor -%}
+};
+
+}
+
+const uint32_t {{bus.name}}_keys[] = {
+    0,
+    {% for msg in bus.frame -%}
+        {{ get_keys(msg, [bus.name, msg.name]) }}
+    {%- endfor -%}
+};
+
+const std::pair<uint32_t, uint32_t> {{bus.name}}_frame_id_range[] = {
+    std::make_pair(1,0),
+    {% for msg in bus.frame -%}
+        {{ get_message_ranges(msg, [bus.name, msg.name]) }}
+    {%- endfor -%}
+};
```

Comment:
Would it be possible to make these things `static` or put them in an unnamed namespace, or otherwise hide them for the user?